File: src/Transcript.java


import java.util.ArrayList;

public class Transcript {
    private StudentID _studentId;
    private ArrayList<TakenCourse> _taken_courses;
    private float _GPA;

    Transcript(StudentID studentId){
        _studentId = studentId;
        _taken_courses = new ArrayList<TakenCourse>();
        _GPA = 0;
    }

    public boolean addTakenCourses(ArrayList<TakenCourse> takenCourses) {
        _taken_courses.addAll(takenCourses);
        calculateGPA();
        return true;
    }

    public boolean addTakenCourse(TakenCourse tc) {
        _taken_courses.add(tc);
        calculateGPA();
        return true;
    }

    private void calculateGPA(){
        _GPA = 0;
        if (_taken_courses.isEmpty()) {
            return;
        }
        for (TakenCourse tc : _taken_courses) {
            float x = switch (tc.get_course_score()) {
                case AA -> 4;
                case BA -> 3.5f;
                case BB -> 3;
                case CB -> 2.5f;
                case CC -> 2;
                case DC -> 1.5f;
                case DD -> 1;
                default -> 0;
            };
            _GPA += x;
        }
        _GPA /= _taken_courses.size();
    }

    //getters

    public ArrayList<TakenCourse> getTakenCourses() {
        return _taken_courses;
    }

    public float get_GPA() {
        calculateGPA();
        return _GPA;
    }
}

File: pythonCode/Transcript.py


import Student
from Course import TakenCourse
import logging

from pythonCode.Logger import setup_logger
logger = setup_logger("Transcript")

class Transcript:
    def __init__(self, student_id):
        logging.getLogger().handlers.clear()
        logger.info(f"{self.__class__.__name__} classes created.")
        self._student_id = student_id
        self._taken_courses = []
        self._gpa = 0.0



    @staticmethod
    def from_dict(data):
        sid = Student.StudentID.from_dict(data["_student_id"])
        tc_list =list()

        for dtc in data["_taken_courses"]:
            tc_list.append(TakenCourse.from_dict(dtc))

        tc = Transcript(sid)
        tc.addTakenCourses(tc_list)
        tc.calculateGPA()
        return tc


    def addTakenCourses(self, taken_courses):
        logger.info("Adding taken courses to the transcript.")
        self._taken_courses.extend(taken_courses)
        self.calculateGPA()

        return True

    def addTakenCourse(self, taken_course):
        logger.info("Adding taken course to the transcript.")
        self._taken_courses.append(taken_course)
        self.calculateGPA()

        return True

    def calculateGPA(self):
        logger.info("Calculating GPA.")
        self._gpa = 0.0
        if not self._taken_courses:

            return

        for course in self._taken_courses:
            score = {
                "AA": 4.0,
                "BA": 3.5,
                "BB": 3.0,
                "CB": 2.5,
                "CC": 2.0,
                "DC": 1.5,
                "DD": 1.0
            }.get(course.getCourseScore(), 0.0)


            self._gpa += score

        self._gpa /= len(self._taken_courses)


    def getTakenCourses(self):
        logger.info(f"The getTakenCourses method in the {self.__class__.__name__} class is called.")
        return self._taken_courses

    def get_GPA(self):
        logger.info(f"The get_GPA method in the {self.__class__.__name__} class is called.")
        self.calculateGPA()
        return self._gpa
